### PR TITLE
refactor(perf): use libp2p-request-response

### PR DESCRIPTION
**Pull request by mxinden**
## Description

Rewrite `libp2p-perf` protocol using `libp2p-request-response`. Additionally adjust to latest conventions, e.g. the final JSON output.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/perf/src/bin/perf-client.rs [hunk 1]
```diff
@@ -43,15 +44,217 @@ async fn main() -> Result<()> {
...
+    ];
+
+    for benchmark in benchmarks {
+        info!(
+            "{}",
+            format!("Start benchmark: {}", benchmark.name()).underline(),
+        );
+
+        benchmark.run(opts.server_address.clone()).await?;
+    }
```
Comment (thomaseizinger) [inline protocols/perf/src/bin/perf-client.rs]: I find this indirection confusing. Why not just have 3 free-functions and call them in a row? That would remove the need for a trait, a loop and the declaration of the array. Seems a lot simpler to me.

```suggestion
benchmark_throughput(opts.server_address.clone()).await?;
benchmark_requests_per_second(opts.server_address.clone()).await?;
benchmark_connections_per_second(opts.server_address).await?;
```

**Comment by mxinden** [comment]
Sample output:

```
➜  perf git:(perf) cargo run --bin perf-client -- --server-address /dns4/libp2p-perf.max-inden.de/udp/4001/quic-v1
[2023-03-20T13:50:02Z INFO  perf_client] Start benchmark: round-trip-time latency
[2023-03-20T13:50:38Z INFO  perf_client] Finished: 1000 pings
[2023-03-20T13:50:38Z INFO  perf_client] - 0.0325 s median
[2023-03-20T13:50:38Z INFO  perf_client] - 0.0419 s 95th percentile
    
[2023-03-20T13:50:38Z INFO  perf_client] Start benchmark: single connection single channel throughput
[2023-03-20T13:50:48Z INFO  perf_client] Finished: sent 10.00 MiB in 7.38 s and received 10.00 MiB in 3.21 s
[2023-03-20T13:50:48Z INFO  perf_client] - 10.84 MiBit/s up
[2023-03-20T13:50:48Z INFO  perf_client] - 24.90 MiBit/s down
    
[2023-03-20T13:50:48Z INFO  perf_client] Start benchmark: single connection parallel requests per second
[2023-03-20T13:50:54Z INFO  perf_client] Finished: sent 10000 1 bytes requests with 1 bytes response each within 5.54 s
[2023-03-20T13:50:54Z INFO  perf_client] - 1803.94 req/s
    
[2023-03-20T13:50:54Z INFO  perf_client] Start benchmark: sequential connections with single request per second
[2023-03-20T13:51:01Z INFO  perf_client] Finished: established 100 connections with one 1 bytes request and one 1 bytes response within 7.23 s
[2023-03-20T13:51:01Z INFO  perf_client] - 0.0384 s 95th percentile connection establishment
[2023-03-20T13:51:01Z INFO  perf_client] - 0.0752 s 95th percentile connection establishment + one request

File: protocols/perf/src/bin/perf-client.rs (hunk 1 above)
Comment (mxinden) [reply protocols/perf/src/bin/perf-client.rs]: `Benchmark` indirection is removed with [fa6886c](https://github.com/libp2p/rust-libp2p/pull/3646/commits/fa6886cd73d09b9b0ff0b7ce6334a42d760ddf88).

**Comment by mxinden** [comment]
@thomaseizinger can you give this pull request a review? Note that with https://github.com/libp2p/rust-libp2p/pull/3990 we can easily do (breaking) changes in the future. Given that and given that it is not a production-environment component, I don't think it requires a thorough review.

In my eyes it makes sense to merge at this point in order to not accumulate even more changes into a single pull request. As is, it is a useful component. That said, I do expect more changes to come.

File: protocols/perf/src/bin/perf.rs [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+    .substream_upgrade_protocol_override(upgrade::Version::V1Lazy)
+    .build();
+
+    swarm
+        .listen_on(
+            Multiaddr::empty()
+                .with(server_address.ip().into())
+                .with(Protocol::Tcp(server_address.port())),
+        )
+        .unwrap();
```
Comment (thomaseizinger) [inline protocols/perf/src/bin/perf.rs]:
```suggestion
        )?;
```

File: protocols/perf/src/bin/perf.rs [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+    };
+
+    SwarmBuilder::with_tokio_executor(
+        transport,
+        libp2p_perf::client::Behaviour::default(),
+        local_peer_id,
+    )
+    .substream_upgrade_protocol_override(upgrade::Version::V1Lazy)
+    .build()
+}
```
Comment (thomaseizinger) [inline protocols/perf/src/bin/perf.rs]: I think you could deduplicate this with the server one by passing in a keypair and constructing the `Behaviour` with the `Default` trait.

**Review by thomaseizinger — APPROVED** [review]

File: protocols/perf/src/bin/perf.rs (hunk 3 above)
Comment (mxinden) [reply protocols/perf/src/bin/perf.rs]: Good idea. Done with the latest commit.

**Comment by mxinden** [comment]
Switched from Noise to TLS. That was as simple as going from `libp2p_noise::Config` to `libp2p_tls::Config`. Thanks for pushing all the consistency efforts @thomaseizinger!

**Comment by thomaseizinger** [comment]
> Switched from Noise to TLS. That was as simple as going from `libp2p_noise::Config` to `libp2p_tls::Config`. Thanks for pushing all the consistency efforts @thomaseizinger!

It is coming together!

**Pull request merged**